src/arithmetic/algebraic_expression.c

```diff
@@ -410,56 +389,95 @@ static Node* _SharedNode(const Edge *a, const Edge *b) {
     return NULL;
 }
 
-static AlgebraicExpression* _AlgebraicExpression_From_Path(Edge **path, uint path_len, const AST *ast) {
+static bool _shouldReversePath(QGEdge **path, uint path_len, bool *to_transpose) {
+    if (path_len <= 1) {
+        to_transpose[0] = false;
+        return false;
+    }
+
+    uint transposeCount = 0;
+    // For every edge except the last:
+    for(uint i = 0; i < path_len - 1; i++) {
+        to_transpose[i] = false;
+        QGEdge *e = path[i];
+        QGEdge *follow = path[i+1];
+        QGNode *shared = _SharedNode(e, follow);
+        // The edge should be transposed if its destination is shared
+        if (e->dest != shared) {
+            to_transpose[i] = true;
+            transposeCount++;
+        }
+    }
+    // For the last edge, transpose if its source is shared
+    QGEdge *e = path[path_len-1];
+    QGNode *shared = _SharedNode(path[path_len-2], e);
+    if (e->src != shared) {
+        transposeCount++;
+        to_transpose[path_len-1] = true;
+    } else {
+        to_transpose[path_len-1] = false;
+    }
+    return transposeCount > path_len - transposeCount;
+}
+
+static void _reversePath(QGEdge **path, uint path_len, bool *to_transpose) {
+    for (uint i = 0; i < path_len; i ++) {
+        // A reversed path should have its transpositions flipped
+        to_transpose[i] = !to_transpose[i];
+        if (to_transpose[i]) QGEdge_Reverse(path[i]);
+    }
+    // Reverse the path array as well as the transposition array
```

Comment:
Previously, we'd call `Edge_Reverse` while iterating over a path, and at the end we'd transpose the AlgebraicExpression if the transpose count was too high.

With the switch to `QG` entities, this became inadequate, because `AlgebraicExpression_Transpose` would only fix the src node, dest node, and edge (if referred). The inconsistencies in `QG` entities caused some breaks in the execution plan logic immediately following AlgebraicExpression construction.